src/expertsystem/reaction/topology.py

```diff
@@ -336,42 +339,139 @@ def __eq__(self, other: object) -> bool:
+        return self.__topology.edges
+
+    def get_initial_state_edge_ids(self) -> List[int]:
+        return self.__topology.get_initial_state_edge_ids()
+
+    def get_final_state_edge_ids(self) -> List[int]:
+        return self.__topology.get_final_state_edge_ids()
+
+    def get_intermediate_state_edge_ids(self) -> List[int]:
+        return self.__topology.get_intermediate_state_edge_ids()
+
+    def get_edge_ids_ingoing_to_node(self, node_id: int) -> List[int]:
+        return [
+            edge_id
+            for edge_id, edge in self.__topology.edges.items()
+            if edge.ending_node_id == node_id
+        ]
+
+    def get_edge_ids_outgoing_from_node(self, node_id: int) -> List[int]:
+        return [
+            edge_id
+            for edge_id, edge in self.__topology.edges.items()
+            if edge.originating_node_id == node_id
+        ]
+
+    def get_originating_node_list(self, edge_ids: Iterable[int]) -> List[int]:
+        return self.__topology.get_originating_node_list(edge_ids)
+
+    def get_originating_final_state_edge_ids(self, node_id: int) -> List[int]:
+        fs_edges = self.__topology.get_final_state_edge_ids()
+        edge_list = []
+        temp_edge_list = self.get_edge_ids_outgoing_from_node(node_id)
+        while temp_edge_list:
+            new_temp_edge_list = []
+            for edge_id in temp_edge_list:
+                if edge_id in fs_edges:
+                    edge_list.append(edge_id)
+                else:
+                    new_node_id = self.__topology.edges[edge_id].ending_node_id
+                    if new_node_id is not None:
+                        new_temp_edge_list.extend(
+                            self.get_edge_ids_outgoing_from_node(new_node_id)
+                        )
+            temp_edge_list = new_temp_edge_list
+        return edge_list
+
+    def get_originating_initial_state_edge_ids(
+        self, node_id: int
+    ) -> List[int]:
+        is_edges = self.__topology.get_initial_state_edge_ids()
+        edge_list = []
+        temp_edge_list = self.get_edge_ids_ingoing_to_node(node_id)
+        while temp_edge_list:
+            new_temp_edge_list = []
+            for edge_id in temp_edge_list:
+                if edge_id in is_edges:
+                    edge_list.append(edge_id)
+                else:
+                    new_node_id = self.__topology.edges[
+                        edge_id
+                    ].originating_node_id
+                    if new_node_id is not None:
+                        new_temp_edge_list.extend(
+                            self.get_edge_ids_ingoing_to_node(new_node_id)
+                        )
+            temp_edge_list = new_temp_edge_list
+        return edge_list
+
+    def get_node_props(self, node_id: int) -> InteractionProperties:
+        return self.__node_props[node_id]
+
+    def get_edge_props(self, edge_id: int) -> _EdgeType:
+        return self.__edge_props[edge_id]
+
+    def evolve(
```

Comment:
I like this new method. Might need a docstring here though, because it seems to be an important method.

Reply:
done